src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,352 @@
+            }
+
+            public async Task<ImmutableArray<TypeImportCompletionItem>> GetAccessibleTopLevelTypesFromCompilationReferenceAsync(
+                Solution solution,
+                Compilation compilation,
+                CompilationReference compilationReference,
+                ImmutableHashSet<string> excludedNamespaces,
+                CancellationToken cancellationToken)
+            {
+#if DEBUG
+                DebugObject.IsCurrentCompilation = false;
+#endif
+
+                if (!(compilation.GetAssemblyOrModuleSymbol(compilationReference) is IAssemblySymbol assemblySymbol))
+                {
+                    return ImmutableArray<TypeImportCompletionItem>.Empty;
+                }
+
+                var isInternalsVisible = compilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(assemblySymbol);
+                var assemblyProject = solution.GetProject(assemblySymbol, cancellationToken);
+                var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(assemblyProject, cancellationToken).ConfigureAwait(false);
+
+                return GetAccessibleTopLevelTypesWorker(assemblyProject.Id, assemblySymbol.GlobalNamespace, checksum, excludedNamespaces, isInternalsVisible, _projectItemsCache, cancellationToken);
+            }
+
+            public ImmutableArray<TypeImportCompletionItem> GetAccessibleTopLevelTypesFromPEReference(
+                Solution solution,
+                Compilation compilation,
+                PortableExecutableReference peReference,
+                ImmutableHashSet<string> excludedNamespaces,
+                CancellationToken cancellationToken)
+            {
+                if (!(compilation.GetAssemblyOrModuleSymbol(peReference) is IAssemblySymbol assemblySymbol))
+                {
+                    return ImmutableArray<TypeImportCompletionItem>.Empty;
+                }
+
+                var key = GetReferenceKey(peReference);
+                var isInternalsVisible = compilation.Assembly.IsSameAssemblyOrHasFriendAccessTo(assemblySymbol);
+                var rootNamespaceSymbol = assemblySymbol.GlobalNamespace;
+
+                if (key == null)
+                {
+                    // Can't cache items for reference with null key, so just create them and return. 
+                    return GetCompletionItemsForTopLevelTypeDeclarations(rootNamespaceSymbol, ns => !excludedNamespaces.Contains(ns), isInternalsVisible);
+                }
+
+                var checksum = SymbolTreeInfo.GetMetadataChecksum(solution, peReference, cancellationToken);
+                return GetAccessibleTopLevelTypesWorker(key, rootNamespaceSymbol, checksum, excludedNamespaces, isInternalsVisible, _peItemsCache, cancellationToken);
+
+                static string GetReferenceKey(PortableExecutableReference reference)
+                    => reference.FilePath ?? reference.Display;
```

Comment:
when do we have a PEReference without a filepath?  is that something we need to support?

Reply:
I think this is the case for VB project reference.

Reply:
if you have a project reference, why do we need to go through a PE-reference?  why not just use a CompilationReference?

Reply:
what I meant is VB project references exist in Compilation.References as PE-reference instead of CompilationReference

Reply:
Oh, that semes like we should just go to the project/compilation for that ref in that case right?  those PE refs might not exist, or might be out of date (for example, if there's a compiler issue in the code).

Reply:
Hmm, what about F# project? Which would be a project doesn't support compilation, right? Tehn only way to get the symbols from it would be through PEReference.

Reply:
Now that I switched to using project reference, I will need to test F#.